textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/plangen/OperatorGraph.java

```diff
@@ -0,0 +1,367 @@
+  
+        operatorTypeMap.put(operatorID, operatorType);
+        operatorPropertyMap.put(operatorID, operatorProperties);
+        adjacencyList.put(operatorID, new HashSet<>());
+        
+        IOperator operator = PlanGenUtils.buildOperator(operatorType, operatorProperties);
+        operatorObjectMap.put(operatorID, operator);
+    }
+    
+    /**
+     * Adds a link from "src" operator to "dest" operator in the graph.
+     * 
+     * @param src, the operator ID of src operator
+     * @param dest, the operator ID of dest operator
+     * @throws PlanGenException, if the operator is null, is empty, or doesn't exist.
+     */
+    public void addLink(String src, String dest) throws PlanGenException {
+        PlanGenUtils.planGenAssert(src != null, "src operator is null");
+        PlanGenUtils.planGenAssert(dest != null, "dest operator is null");
+        
+        PlanGenUtils.planGenAssert(! src.trim().isEmpty(), "src operator is empty");
+        PlanGenUtils.planGenAssert(! dest.trim().isEmpty(), "dest operator is empty");
+        
+        PlanGenUtils.planGenAssert(hasOperator(src), String.format("operator %s doesn't exist", src));
+        PlanGenUtils.planGenAssert(hasOperator(dest), String.format("operator %s doesn't exist", dest));
+        
+        adjacencyList.get(src).add(dest);
+    }
+    
+    /**
+     * Returns true if the operator graph contains the operatorID.
+     * 
+     * @param operatorID
+     * @return
+     */
+    public boolean hasOperator(String operatorID) {
+        return adjacencyList.containsKey(operatorID);
+    }
+    
+    /**
+     * Builds and returns the query plan from the operator graph.
+     * 
+     * @return the plan generated from the operator graph
+     * @throws Exception, if the operator graph is invalid.
+     */
+    public Plan buildQueryPlan() throws Exception {
+        validateOperatorGraph();
+        connectOperators();
+        ISink sink = findSinkOperator();
+        
+        Plan queryPlan = new Plan(sink);
+        return queryPlan;
+    }
+
+
+    /*
+     * Validates the operator graph.
+     * The operator graph must meet all of the following requirements:
+     * 
+     *   the graph is a DAG (directed acyclic graph)
+     *     this DAG is weakly connected (no unreachable vertices).
+     *     there's no cycles in this DAG.
+     *   each operator must meet its input and output arity constraints.
+     *   the operator graph has at least one source operator.
+     *   the operator graph has exactly one sink.
+     * 
+     * Throws PlanGenException if the operator graph is invalid.
+     */
+    private void validateOperatorGraph() throws PlanGenException {
+        checkGraphConnectivity();
+        checkGraphCyclicity();
+        checkOperatorInputArity();
+        checkOperatorOutputArity();
+        checkSourceOperator();
+        checkSinkOperator();
+    }
+  
+    /*
+     * Detects if there are any operators not connected to the operator graph.
```

Comment:
This sentence is not accurate.  How about "Detects if the graph can be partitioned disjoint subgraphs"?